### ISSUE TITLE
Bag of Crafting preview

### ISSUE THREAD

**Issue by wofsauge** [wofsauge/External-Item-Descriptions#69]
Algorithm: https://github.com/bladecoding/isaac-bagofcrafting/blob/main/src/BagOfCrafting.ts

**Comment by wofsauge** [comment]
Update: 
This feature is implemented now, but only works with the pickups on the ground.

The Bag of Crafting Content is not accessible with the current API. This needs a workaround or maybe a new API feature with the next game update.

**Comment by amaurymaille** [comment]
I've thought of two similar ways to work around the limitations of the API regarding the content of the BOC, but they are both quite resource intensive.

1. Find the reason a pickup is no longer in the room. In order to do that, bind callbacks to every single action that can affect pickups (which would include using the D20, the D100, getting hit and triggering Cracked Dice, using the Dice Shard, the Black Rune, the Blank Rune triggering Black Rune, using Moving Box, Compost (I think ?), Bumbino picking up a coin (kill me) and so on) so we can know why a pickup disappeared from a room. Each room is given an array of pairs (pickup, state), every time a pickup spawns add it to that array and flag it as "ground". Callbacks on effects that may affect pickups will update the state accordingly. Meanwhile, in the POST_UPDATE callback, compare the pickups in the array associated with the current room with the pickups on the ground: if a pickup has not state "ground", remove it from the array; if a pickup has state "ground" but is no longer in the room, it has been put in the bag of crafting. Add that pickup to a circular buffer representing the BOC.
2. Find the reason a pickup has been removed. This is similar to the previous one, except we replace POST_UPDATE with POST_ENTITY_REMOVE which may be less intensive than every frame update ? This one however comes with the drawback that it is called for each room transition because the game effectively unloads the pickups (MC_POST_NEW_ROOM to the rescue ?).

We also need to take into account the fact that the player may cycle between pickups inside the boc, so the input used to cycle must perform a shift in the circular buffer. I just checked it, apparently pickups keep cycling in the boc even if you use Ctrl / whatever on a controller to switch card / rune / pill, so at least we do not need to consider whether the player has switched to pill / card / rune when processing a use of the cycle button.

That's the most straightforward way I've thought of doing it. Maybe we'll get a GetBagOfCraftingContent() in an update, though I find that unlikely given that we already have functions for things like Soul / Blood Charges. Managing Bag of Crafting would have been almost trivial to add.

I can give it a try if you think it may be worth checking it

**Comment by wofsauge** [comment]
all the things you mentioned are very resource intensive indeed.  The best way i know to see if something got picked up by the bag is by checking the collision with the bag entity (8.4.4) as seen here in this screenshot.
![image](https://user-images.githubusercontent.com/245725/119894091-77930980-bf3c-11eb-824e-74dc7d83394b.png)

If we just keep trak of all the pickups collided with that entity, we should have a list of bag content.

As you mention tho, the next problem is the cycling. this can be easily be tracked with the action buttons. The other thing of course is the crafting action.

other than that, it should work i guess.

Its still pretty hacky so i really hope for a future update that allows us a look into the bag content.

**Comment by amaurymaille** [comment]
Interestingly, entity type 8 corresponds to Mom's Knife (as a quick aside, I almost expected Forgotten's Bone to be a subtype of ENTITY_KNIFE because of how similarly it behaves to the Bag of Crafting and... It's apparently not). As another aside, how did you display the entities data ? I checked the debug flags in the console but none of them seems to actually display these infos.

Here is what I found. While Isaac is holding the bag over his head, no updates are performed. Once the bag is swung in a direction, MC_PRE_KNIFE_UPDATE is called for every frame during which the bag entity exists. However there is a catch: neither MC_PRE_KNIFE_COLLISION nor MC_PRE_PICKUP_COLLISION are called when the bag hits something, be it an enemy or a pickup. I attached a function to every single MC_*_COLLISION to see if the BOC behaved like a different entity when swung, but apparently not... No callbacks was triggered with the bag-knife as parameter.

As you suggested, checking the collision manually seems the way to go. Here are three screenshots showcasing the hitspheres of the bag (yes, there are multiple ones). First screenshot is for Cain's hitsphere reference. I used the debug console to display the hitspheres.

![20210528115801_1](https://user-images.githubusercontent.com/2423678/119967017-36d2d900-bfac-11eb-93a2-d622c8c184db.jpg)
![20210528115806_1](https://user-images.githubusercontent.com/2423678/119966974-27539000-bfac-11eb-8f7e-7ca957932145.jpg)
![20210528115824_1](https://user-images.githubusercontent.com/2423678/119966976-27ec2680-bfac-11eb-9749-9d9025c0c1e6.jpg)

This allows to demystify how the bag works with the Range stat. There are three additional hitspheres when the bag is swung: one red around the player that is slightly bigger and *unaffected* by Range, one red that is farther away from the player and is affected by Range, and a yellow one that is dependent on the second red one (although it remains at the same distance from it). After several verifications, the second red hitsphere and the yellow hitsphere are the only ones that are used to pick pickups up and to damage enemies and they are extremely precise (I spawned the Sub Horfs from Downpour to check the precision of the hitspheres, as well as coins). So even if there is a pickup right in front of the player, if the Range stat is high enough, that pickup won't be picked by the bag. Thanks Ed.

I don't have any idea why there is another slightly bigger hitsphere around the player when the bag is swung however. Given that is corresponds to a knife entity, I don't think it's related to the player.

Back to Lua. The yellow hitsphere isn't associated with a knife entity, but both red hitspheres are. The first one is entity 8.4.0, the second one (which is of interest) is entity 8.4.4 (8.4.0 is hidden behind 8.4.4 in your screenshot). If I'm not mistaken, if we manage to either find the entity that corresponds to the yellow hitsphere, or if we can just deduce the distance between the red and yellow hitspheres as well as the radius of the yellow hitsphere, we can deduce if a pickup was put in the bag. 

Going this way, I think we would need to combine both the MC_POST_KNIFE_UPDATE callback with the MC_PRE_ENTITY_REMOVE callback, because we can't be sure of the order in which the game will actually remove the pickups from the ground and put them in the bag (what if there are multiple pickups stacked on top of each other ?). An easy way would be to tag pickups hit by the bag during the MC_POST_KNIFE_UPDATE event with something so we could check in MC_PRE_ENTITY_REMOVE if the entity was put in the bag or not.

And then of course we need to intercept the use of the bag to clear the bag's content. MC_POST_USE_ITEM should be the way to go I think ?
